server/src/types/assignment.ts

```diff
@@ -0,0 +1,52 @@
+// SPDX-FileCopyrightText: 2023 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+export interface AssignmentParams {
+  min: number;
+  max: number;
+}
+export type ManuallyGradedParams = {
+  min: number;
+  max: number;
+}
+
+export abstract class Assignment<P extends AssignmentParams> {
+  public abstract calculateGrade(
+    subGrades:
+      this extends ManuallyGradedAssignment ? number: Array<number>
```

Comment:
Instead of `Array<number>`, how about:
```typescript
class Attainment<P extends AssignmentParams> {
    public assignment: Assignment<P>;
    public subAttainments: Array<Attainment<???> | number>;
}
```
But I'm unsure about what to put in place of `???` for the generic params.